pulsar-log4j2-appender/src/main/java/org/apache/pulsar/log4j2/appender/PulsarManager.java

```diff
@@ -0,0 +1,135 @@
+
+    static Supplier<ClientBuilder> PULSAR_CLIENT_BUILDER = () -> PulsarClient.builder();
+
+    static BiFunction<String, byte[], Message<byte[]>> MESSAGE_BUILDER = (key, data) -> {
+        MessageBuilder<byte[]> messageBuilder = MessageBuilder.create()
+            .setContent(data);
+        if (null != key) {
+            messageBuilder = messageBuilder.setKey(key);
+        }
+        return messageBuilder.build();
+    };
+
+    private PulsarClient client;
+    private Producer<byte[]> producer;
+
+    private final String serviceUrl;
+    private final String topic;
+    private final String key;
+    private final boolean syncSend;
+
+    public PulsarManager(final LoggerContext loggerContext,
+                         final String name,
+                         final String serviceUrl,
+                         final String topic,
+                         final boolean syncSend,
+                         final Property[] properties,
+                         final String key) {
+        super(loggerContext, name);
+        this.serviceUrl = Objects.requireNonNull(serviceUrl, "serviceUrl");
+        this.topic = Objects.requireNonNull(topic, "topic");
+        this.syncSend = syncSend;
+        this.key = key;
+    }
+
+    @Override
+    public boolean releaseSub(final long timeout, final TimeUnit timeUnit) {
+        if (producer != null) {
+            try {
+                producer.close();
+            } catch (PulsarClientException e) {
+                // exceptions on closing
+            }
+        }
+        return true;
+    }
+
+    public void send(final byte[] msg)  {
+        if (producer != null) {
+            String newKey = null;
+
+            if(key != null && key.contains("${")) {
+                newKey = getLoggerContext().getConfiguration().getStrSubstitutor().replace(key);
+            } else if (key != null) {
+                newKey = key;
+            }
+
+            Message<byte[]> message = MESSAGE_BUILDER.apply(newKey, msg);
+            if (syncSend) {
+                try {
+                    producer.send(message);
+                } catch (PulsarClientException e) {
+                    LOGGER.error("Unable to write to Pulsar in appender [" + getName() + "]", e);
+                }
+            } else {
+                producer.sendAsync(message)
+                    .exceptionally(cause -> {
+                        LOGGER.error("Unable to write to Pulsar in appender [" + getName() + "]", cause);
+                        return null;
+                    });
+            }
+        }
+    }
+
+    public void startup() throws Exception {
+        try {
+            client = PULSAR_CLIENT_BUILDER.get()
+                .serviceUrl(serviceUrl)
+                .build();
+            producer = client.newProducer()
+                .topic(topic)
+                .producerName("pulsar-log4j2-appender-" + topic)
+                .enableBatching(true)
```

Comment:
If the mode is `syncSend`, batching would just increase the latency (due to waiting for time budget) and decrease the throughput, since we only have 1 in-flight message.